plugin/vfs/plugin.go

```diff
@@ -0,0 +1,127 @@
+package vfs
+
+import (
+	"bytes"
+	"context"
+	"io/fs"
+	protoapi "sigs.k8s.io/kube-scheduler-wasm-extension/kubernetes/proto/api"
+	"strconv"
+	"sync/atomic"
+	"testing/fstest"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	"sigs.k8s.io/kube-scheduler-wasm-extension/plugin/internal"
+
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/sys"
+)
+
+func IsVFSPlugin(module wazero.CompiledModule) bool {
+	return internal.DetectImports(module.ImportedFunctions()) == internal.ModeWasiP1
+}
+
+func NewPlugin(runtime wazero.Runtime, module wazero.CompiledModule, guestName string) (pl framework.Plugin, err error) {
+	return &vfsPlugin{
+		moduleConfig:    wazero.NewModuleConfig(),
+		guestName:       guestName,
+		runtime:         runtime,
+		module:          module,
+		instanceCounter: atomic.Uint64{},
+	}, nil
+}
+
+type vfsPlugin struct {
+	moduleConfig wazero.ModuleConfig
+	guestName    string
+
+	runtime wazero.Runtime
+	module  wazero.CompiledModule
+
+	instanceCounter atomic.Uint64
+}
+
+var _ framework.FilterPlugin = (*vfsPlugin)(nil)
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (pl *vfsPlugin) Name() string {
+	return internal.PluginName
+}
+
+var _ fs.FS = (*vfs)(nil)
+
+// vfs is a virtual file system which allows use of all pod and nodeInfo data
+// without custom ABIs. The caller can deserialize the top-level proto, or
+// more specific ones.
+type vfs struct {
+	pod      *v1.Pod
+	nodeInfo *framework.NodeInfo
+}
+
+func (v vfs) Open(name string) (fs.File, error) {
+	var marshaller func() ([]byte, error)
+	switch name {
+	case "pod/spec":
+		// TODO v.pod.Spec.Marshal is incompatible, find a way to automatically
+		// convert *v1.PodSpec to protoapi.IoK8SApiCoreV1PodSpec
```

Comment:
this part is still an issue, we at least need a better comment as to why these are different encoding (e.g. what encoding `v1.PodSpec.Marshal` is and why it isn't the same as `protoapi.IoK8SApiCoreV1PodSpec.MarshalVT`